src/unix/epoll.rs

```diff
@@ -0,0 +1,429 @@
+        /// descriptors to receive an event.
+        const EXCLUSIVE = EPOLLEXCLUSIVE as u32;
+    }
+}
+
+/// Wrapper over 'libc::epoll_event'.
+///
+// We are using `transparent` here to be super sure that this struct and its fields
+// have the same alignment as those from the `epoll_event` struct from C.
+#[repr(transparent)]
+#[derive(Clone)]
+pub struct EpollEvent(epoll_event);
+
+impl Deref for EpollEvent {
+    type Target = epoll_event;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Default for EpollEvent {
+    fn default() -> Self {
+        EpollEvent(epoll_event {
+            events: 0u32,
+            u64: 0u64,
+        })
+    }
+}
+
+impl EpollEvent {
+    /// Create a new epoll_event instance with the following fields: `events`, which contains
+    /// an event mask and `data` which represents a user data variable. `data` field can be
+    /// a fd on which we want to monitor the events specified by `events`.
+    pub fn new(events: EventSet, data: u64) -> Self {
+        EpollEvent(epoll_event {
+            events: events.bits(),
+            u64: data,
+        })
+    }
+
+    /// Returns the `events` from `libc::epoll_event`.
+    pub fn events(&self) -> u32 {
+        self.events
+    }
+
+    /// Returns the `EventSet` corresponding to `epoll_event.events`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `libc::epoll_event` contains invalid events.
+    pub fn event_set(&self) -> EventSet {
+        // This unwrap is safe because `epoll_events` can only be user created or
+        // initialized by the kernel. We trust the kernel to only send us valid
+        // events. The user can only initialize `epoll_events` using valid events.
+        EventSet::from_bits(self.events()).unwrap()
+    }
+
+    /// Returns the `data` from the `libc::epoll_event`.
+    pub fn data(&self) -> u64 {
+        self.u64
+    }
+
+    /// Converts the `libc::epoll_event` data to a RawFd.
+    ///
+    /// This conversion is lossy when the data does not correspond to a RawFd
+    /// (data does not fit in a i32).
+    pub fn fd(&self) -> RawFd {
+        self.u64 as i32
+    }
+}
+
+/// Wrapper over epoll functionality.
+#[derive(Debug)]
+pub struct Epoll {
+    epoll_fd: RawFd,
+}
+
+impl Epoll {
+    /// Create a new epoll file descriptor.
+    pub fn new() -> io::Result<Self> {
+        let epoll_fd = SyscallReturnCode(unsafe { epoll_create1(EPOLL_CLOEXEC) }).into_result()?;
+        Ok(Epoll { epoll_fd })
+    }
+
+    /// Wrapper for `libc::epoll_ctl`.
+    ///
+    /// This can be used for adding, modifying or removing a file descriptor in the
+    /// interest list of the epoll instance.
+    ///
+    /// # Arguments
+    ///
+    /// * `operation` refers to the action to be performed on the file descriptor.
+    /// * `fd` is the file descriptor on which we want to perform `operation`.
+    /// * `event` refers to the `epoll_event` instance that is linked to `fd`.
+    pub fn ctl(
```

Comment:
Can you add `/// # Examples` sections to the public facing APIs too?

Reply:
Done.